gnutls bugfix: Fix error handling in gtlsRecordRecv

alorbach opened the pull request:
There was a rare possibility that the E_AGAIN/E_INTERRUPT handling
could cause an infinite loop (100% CPU Usage), for example when a TLS
handshake is interrupted at a certain stage.

- After gnutls_record_recv is called, and E_AGAIN/E_INTERRUPT error
  occurs, we need to do additional read/write direction handling
  with gnutls_record_get_direction.
- After the second call of gnutls_record_recv (Expand buffer)
  we needed to also check the eror codes for E_AGAIN/E_INTERRUPT
  to do propper errorhandling.


see also: https://github.com/rsyslog/rsyslog/issues/4818

<!--
LEGAL GDPR NOTICE:
According to the European data protection laws (GDPR), we would like to make you
aware that contributing to rsyslog via git will permanently store the
name and email address you provide as well as the actual commit and the
time and date you made it inside git's version history. This is inevitable,
because it is a main feature git. If you are concerned about your
privacy, we strongly recommend to use

--author "anonymous <gdpr@example.com>"

together with your commit. Also please do NOT sign your commit in this case,
as that potentially could lead back to you. Please note that if you use your
real identity, the GDPR grants you the right to have this information removed
later. However, we have valid reasons why we cannot remove that information
later on. The reasons are:

* this would break git history and make future merges unworkable
* the rsyslog projects has legitimate interest to keep a permanent record of the
  contributor identity, once given, for
  - copyright verification
  - being able to provide proof should a malicious commit be made

Please also note that your commit is public and as such will potentially be
processed by many third-parties. Git's distributed nature makes it impossible
to track where exactly your commit, and thus your personal data, will be stored
and be processed. If you would not like to accept this risk, please do either
commit anonymously or refrain from contributing to the rsyslog project.
-->

lonicerae commented on runtime/nsd_gtls.c:
```diff
@@ -576,14 +577,31 @@ gtlsRecordRecv(nsd_gtls_t *pThis)
+sslerragain:
+		/* Check if the underlaying file descriptor needs to read or write data!*/
+		wantsWriteData = gnutls_record_get_direction(pThis->sess);
+		if (wantsWriteData == 0) {
```
In our case, the problem is the thread keeps doing 'recvfrom()' endlessly. Could we abort the connection after some retries or some time?

alorbach replied:
@rgerhards what do you think? In order to maintain existing 

> In our case, the problem is the thread keeps doing 'recvfrom()' endlessly. Could we abort the connection after some retries or some time?

Could you post a gdb stack trace?
Or even better could you enable debug output, if we could catch that error with debug log enabled that would be awesome.

opoplawski commented:
I have debug output, but it's lot of data and I don't want to post it publicly.

opoplawski commented:
I'm afraid the debug output is 1.5GB compressed and too large for that form to upload.

alorbach commented:
> I'm afraid the debug output is 1.5GB compressed and too large for that form to upload.
 I have extended the form to support up to 10GB files. Can you retry?

alorbach commented:
> I've uploaded the segment. Hope it helps. Sorry for the delay - vacation and some technical issues.

thanks I got it 👍 , I will review the logfile ASAP.

alorbach commented:
> I've uploaded the segment. Hope it helps. Sorry for the delay - vacation and some technical issues.
I reviewed the debug logfile and this really looks odd. Something is not handeled right at some point, properly with a huge amount of sessions. 

I have added some more debug output and added output of the nsd pointer at more debug outputs so we can track which nsd is doing what. For some reason it never reaches the point of "nsdsel_gtls: reached end".  

@opoplawski  Can you try to run with the amended debug output changes from here? It will help me better tracking the nsd objects inside the debug logfile.

opoplawski commented:
I've started running it, but it's generating about 1GB of output each minute - so not sure how long I can run it.

opoplawski commented:
New file uploaded

alorbach commented:
@opoplawski Thanks a lot we are coming closer ;)

> tcpsrv.c: tcpsrv: ready to process 1 event entries
> imtcp.c        : tcpsrv.c: tcpsrv: processing item 15, pUsr 0x7f133c1bbb60, bAbortConn
> imtcp.c        : tcpsrv.c: netstream 0x7f133c1bba30 with new data
> imtcp.c        : nsd_gtls.c: gtlsRecordRecv: start
> imtcp.c        : nsd_gtls.c: gtlsRecordRecv: gnutls_record_recv received 456 bytes
> imtcp.c        : nsd_gtls.c: gtlsRecordRecv return. nsd 0x7f133c19ab60, iRet 0, lenRcvd 456, lenRcvBuf 456, ptrRcvBuf 0
> imtcp.c        : nsd_gtls.c: gtlsRcv return. nsd 0x7f133c19ab60, iRet 0, lenRcvBuf -1, ptrRcvBuf 456

The same tcp session is causing the loop over and over, however I can't see why. It looks like rsyslog is getting hammered with the same amount of data on that TCP Session over and over again. 

I have added more debug output, it will now reveal the tcp session client, causing the loop. Its ok to have debug log run for 1 minute, I don't need more data. 

Thanks a lot for helping.

opoplawski commented:
Uploaded new log.  This is just the final minute before the server was shutdown but probably doesn't capture the initial start of the issue.

opoplawski commented:
@alorbach Is this ready for testing?

alorbach commented:
@opoplawski : Yes I added a potential fix for the problem. However I cannot test it, if you could run it at your end it would be great. When you debug you may see "doRetry: gtlsRecordRecv returned RETRY, but there is no pending" from time to time, when this happens it could have caused the 100% loop before.

alorbach commented:
@opoplawski Did you have a chance to test the PR again?

opoplawski commented:
Yes - it's been running for nearly 2 days and I haven't seen the problem again - which is the longest I've ever seen it go.  So I think this has likely fixed the issue for me.

alorbach commented:
@opoplawski Thanks a lot for the feedback 👍 ! If it hasn't happened again, it most likely is fixed now. 
Can you check if any rsyslog tls errors werre logged during meantime?

alorbach commented:
So it appears that this PR should close the following issues: 
https://github.com/rsyslog/rsyslog/issues/4834
https://github.com/rsyslog/rsyslog/issues/4818
https://github.com/rsyslog/rsyslog/issues/4638
